Recreate new installation after deletion from keychain

cbaker6 opened the pull request:
Automatically create a new installation after it's deleted from the Keychain. A deletion typically occurs on user logout. Fix #110 

- [x] add changelog entry
- [x] add tests

cbaker6 commented:
@funkenstrahlen can you try out this branch and see if you run into the same issue?

The testcases pass and new installationIds are created

funkenstrahlen commented on Tests/ParseSwiftTests/ParseUserTests.swift:
```diff
@@ -1083,59 +1083,75 @@ class ParseUserTests: XCTestCase { // swiftlint:disable:this type_body_length
+                        #if !os(Linux) && !os(Android)
+                        if let installationFromKeychain: CurrentInstallationContainer<BaseParseInstallation>
+                            = try? KeychainStore.shared.get(valueFor: ParseStorage.Keys.currentInstallation) {
+                            if installationFromKeychain.installationId == oldInstallationId {
```
What do you think about extending the test case here and also checking if `installationFromKeychain.currentInstallation != nil`?

cbaker6 replied:
This may not always be the case by the time the test gets here because of the dispatching to main queue. If it is nil when it gets here, it won't be in a couple of seconds

funkenstrahlen replied:
You're right. I haven't thought about the dispatch. The additional test case is not required. Has just been a thought as it's not checked right now.

cbaker6 replied:
#113 should provide more predictable behavior. Let me know your thoughts on that thread

funkenstrahlen commented on Tests/ParseSwiftTests/ParseUserTests.swift:
```diff
@@ -1083,59 +1083,75 @@ class ParseUserTests: XCTestCase { // swiftlint:disable:this type_body_length
+                    DispatchQueue.main.async {
+                        if let installationFromMemory: CurrentInstallationContainer<BaseParseInstallation>
+                            = try? ParseStorage.shared.get(valueFor: ParseStorage.Keys.currentInstallation) {
+                            if installationFromMemory.installationId == oldInstallationId {
```
What do you think about extending the test case here and also checking if `installationFromKeychain.currentInstallation != nil`?

funkenstrahlen commented on Tests/ParseSwiftTests/ParseUserTests.swift:
```diff
@@ -1083,59 +1083,75 @@ class ParseUserTests: XCTestCase { // swiftlint:disable:this type_body_length
+                }
+                DispatchQueue.main.async {
+                    if let installationFromKeychain = BaseParseInstallation.current {
+                        if installationFromKeychain.installationId == oldInstallationId {
```
What do you think about extending the test case here and also checking if `installationFromKeychain.currentInstallation != nil`?

funkenstrahlen commented on Tests/ParseSwiftTests/ParseUserCombineTests.swift:
```diff
@@ -293,32 +293,42 @@ class ParseUserCombineTests: XCTestCase { // swiftlint:disable:this type_body_le
+                    #if !os(Linux) && !os(Android)
+                    if let installationFromKeychain: CurrentInstallationContainer<BaseParseInstallation>
+                        = try? KeychainStore.shared.get(valueFor: ParseStorage.Keys.currentInstallation) {
+                        if installationFromKeychain.installationId == oldInstallationId {
```
What do you think about extending the test case here and also checking if `installationFromKeychain.currentInstallation != nil`?

funkenstrahlen commented on Tests/ParseSwiftTests/ParseUserCombineTests.swift:
```diff
@@ -293,32 +293,42 @@ class ParseUserCombineTests: XCTestCase { // swiftlint:disable:this type_body_le
+                DispatchQueue.main.async {
+                    if let installationFromMemory: CurrentInstallationContainer<BaseParseInstallation>
+                        = try? ParseStorage.shared.get(valueFor: ParseStorage.Keys.currentInstallation) {
+                        if installationFromMemory.installationId == oldInstallationId {
```
What do you think about extending the test case here and also checking if `installationFromKeychain.currentInstallation != nil`?

funkenstrahlen commented:
Thank you for quick response and the PR! I added some suggestions above to have a more detailed test suite.

I added this branch to my project and can confirm it works a lot better now! Thank you!

However I still need to manually save the newly created installation to the server after logout because every subsequent `ParseInstallation.fetch` fails with `ParseError code=-1 error=Cannot fetch an object without id` before it has been saved to the server once.

```swift
    func logout() {
        PodliveUser.logout { (result) in
            switch result {
            case .failure(let error):
                print("Could not logout: " + error.localizedDescription)
            case .success():
                print("successfully logged out")
                NotificationCenter.default.post(name: Notification.Name.userUpdated, object: nil)
                PodliveInstallation.current?.save(completion: { (result) in
                    switch result {
                    case .success(_):
                        self.loginAsAnonymousUser()
                    case .failure(let error):
                        print("Could not save installation: " + error.localizedDescription)
                    }
                })
            }
        }
    }
```

Do you think saving the installation to the server should be done by the framework after logout?

cbaker6 commented:
> Do you think saving the installation to the server should be done by the framework after logout?

I believe this comment may answer this question: https://github.com/parse-community/Parse-Swift/issues/93#issuecomment-798935361

cbaker6 commented:
> 

> Thank you for quick response and the PR! I added some suggestions above to have a more detailed test suite.
> 
> I added this branch to my project and can confirm it works a lot better now! Thank you!
> 
> However I still need to manually save the newly created installation to the server after logout because every subsequent `ParseInstallation.fetch` fails with `ParseError code=-1 error=Cannot fetch an object without id` before it has been saved to the server once.
> 
> ```swift
>     func logout() {
>         PodliveUser.logout { (result) in
>             switch result {
>             case .failure(let error):
>                 print("Could not logout: " + error.localizedDescription)
>             case .success():
>                 print("successfully logged out")
>                 NotificationCenter.default.post(name: Notification.Name.userUpdated, object: nil)
>                 PodliveInstallation.current?.save(completion: { (result) in
>                     switch result {
>                     case .success(_):
>                         self.loginAsAnonymousUser()
>                     case .failure(let error):
>                         print("Could not save installation: " + error.localizedDescription)
>                     }
>                 })
>             }
>         }
>     }
> ```
> 
> Do you think saving the installation to the server should be done by the framework after logout?

Since you logout/login immediately, I can see the possibility of odd behavior because of the need for Installation to be on the main queue. There can be times your anonymous login occurs before the new installation is created. You can minimize this odd behavior by logging in anonymously after a short time, say a second (dispatch after a second). The changes discussed in #113 should help with this also and you might not need to wait anytime after to log back in, but with the current implementation, you should wait a second or ask the user to press a button to login

funkenstrahlen commented:
> > Do you think saving the installation to the server should be done by the framework after logout?
> 
> I believe this comment may answer this question: [#93 (comment)](https://github.com/parse-community/Parse-Swift/issues/93#issuecomment-798935361)

Thank you for linking the related issue thread. Actually I do have custom properties in my Installation object and therefore need to retrieve it manually. However I think you are absolutely correct that fetching the Installation at this point should not be part of the framework.

funkenstrahlen commented:
> Since you logout/login immediately, I can see the possibility of odd behavior because of the need for Installation to be on the main queue. There can be times your anonymous login occurs before the new installation is created. You can minimize this odd behavior by logging in anonymously after a short time, say a second (dispatch after a second).

That's correct. Great you thought about that! I think implementing #113 is the best approach to create a more predictable behavior.

cbaker6 commented:
> That's correct. Great you thought about that! I think implementing #113 is the best approach to create a more predictable behavior.

Even with the changes, you should probably implement one of my suggestions I mentioned above. You can see that because of threading the new installation may not be ready immediately in the keychain: https://github.com/parse-community/Parse-Swift/blob/b671243fa650672af182f0e36e0d947a38d8f3f9/Tests/ParseSwiftTests/ParseUserCombineTests.swift#L313-L336
